### PR TITLE
[JENKINS-36553] Added an option to keep file permissions

### PR DESCRIPTION
**Pull request by Paullo612**
Keep local file permissions of transferred files if configured to do so.

### CODE REVIEW

**Comment by yantimirov-timur** [comment]
Now PR ready to merge.

**Comment by Paullo612** [comment]
@yantimirov-timur, I think you'd better rebase than introducing a merge commit.

Also, looks like build fails with your changes.

**Comment by eugenyk** [comment]
> @yantimirov-timur, I think you'd better rebase than introducing a merge commit.
> 
> Also, looks like build fails with your changes.

@Paullo612 , build fail also in the master branch, thus it's not this PR fault

**Comment by gmcdonald** [comment]
Running 'mvn test' passes locally for me on branch master.

Running 'mvn test' fails locally for me on a branch containing your PR#34 :-

[INFO] Running jenkins.plugins.publish_over_ssh.BapSshClientTest
[ERROR] Tests run: 28, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 6.624 s <<< FAILURE! - in jenkins.plugins.publish_over_ssh.BapSshClientTest
[ERROR] jenkins.plugins.publish_over_ssh.BapSshClientTest.testTransferFile  Time elapsed: 0.014 s  <<< FAILURE!
java.lang.AssertionError: 

  Expectation failure on verify:
    put(EasyMock for class java.io.InputStream, "my.file"): expected: 1, actual: 1
    isKeepFilePermissions(): expected: 1, actual: 0
	at org.easymock.internal.MocksControl.verify(MocksControl.java:111)
	at jenkins.plugins.publish_over_ssh.BapSshClientTest.testTransferFile(BapSshClientTest.java:165)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:365)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:273)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:159)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:377)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:138)
	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:465)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:451)

**Comment by eugenyk** [comment]
> Running 'mvn test' passes locally for me on branch master.
> 
> Running 'mvn test' fails locally for me on a branch containing your PR#34 :-
> 
> [INFO] Running jenkins.plugins.publish_over_ssh.BapSshClientTest [ERROR] Tests run: 28, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 6.624 s <<< FAILURE! - in jenkins.plugins.publish_over_ssh.BapSshClientTest [ERROR] jenkins.plugins.publish_over_ssh.BapSshClientTest.testTransferFile Time elapsed: 0.014 s <<< FAILURE! java.lang.AssertionError:
> 
> Expectation failure on verify: put(EasyMock for class java.io.InputStream, "my.file"): expected: 1, actual: 1 isKeepFilePermissions(): expected: 1, actual: 0 at org.easymock.internal.MocksControl.verify(MocksControl.java:111) at jenkins.plugins.publish_over_ssh.BapSshClientTest.testTransferFile(BapSshClientTest.java:165) at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) at java.base/java.lang.reflect.Method.invoke(Method.java:566) at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59) at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12) at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56) at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17) at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26) at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306) at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100) at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366) at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103) at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63) at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331) at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79) at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329) at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66) at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293) at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26) at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27) at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306) at org.junit.runners.ParentRunner.run(ParentRunner.java:413) at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:365) at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:273) at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238) at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:159) at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:377) at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:138) at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:465) at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:451)

@gmcdonald , test fixed

File: src/main/java/jenkins/plugins/publish_over_ssh/BapSshClient.java
Comment (eugenyk) [inline src/main/java/jenkins/plugins/publish_over_ssh/BapSshClient.java:40]: It's strange to import all by * and later male a particular import. In general it's netter to make particular imports

File: src/main/java/jenkins/plugins/publish_over_ssh/BapSshClient.java
Comment (eugenyk) [inline src/main/java/jenkins/plugins/publish_over_ssh/BapSshClient.java:196]: If variable is immutable after assignment - `final` keyword can be used in order to increase readability

File: src/main/java/jenkins/plugins/publish_over_ssh/BapSshClient.java
Comment (eugenyk) [inline src/main/java/jenkins/plugins/publish_over_ssh/BapSshClient.java:195]: What can cause `InterruptedException`?

File: src/main/java/jenkins/plugins/publish_over_ssh/BapSshClient.java
Comment (eugenyk) [inline src/main/java/jenkins/plugins/publish_over_ssh/BapSshClient.java:204]: Variable declaration is outside the minimal scope, where it is used. According to best practices it's better to declare variable in the closest scope

File: src/main/java/jenkins/plugins/publish_over_ssh/BapSshClient.java
Comment (eugenyk) [inline src/main/java/jenkins/plugins/publish_over_ssh/BapSshClient.java:203]: Is it logically correct to change mode in this method? Maybe it'll be duplicated several times for one directory. Maybe there is some closer place for directory creation? `changeDirectory` or something else

**Review by eugenyk — CHANGES_REQUESTED** [review]

File: src/main/java/jenkins/plugins/publish_over_ssh/BapSshClient.java
Comment (yantimirov-timur) [reply src/main/java/jenkins/plugins/publish_over_ssh/BapSshClient.java:195]: methods `mode()` and `absolutize()`

File: src/main/java/jenkins/plugins/publish_over_ssh/BapSshClient.java
Comment (yantimirov-timur) [reply src/main/java/jenkins/plugins/publish_over_ssh/BapSshClient.java:204]: Without declare ` FilePath parentFle = filePath.getParent()` before null check, spotbug fails with `NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE`

**Comment by gmcdonald** [comment]
Things are looking better, though somehow I have a merge conflict in my local copy now, will sort that out and re-test.
Let me know when you are done tweaking things so I can do a final test then merge.

**Comment by yantimirov-timur** [comment]
@gmcdonald Now you can test and merge

**Comment by gmcdonald** [comment]
> @gmcdonald Now you can test and merge

Thanks, I checked out a fresh branch and it seems we still have an error with 'mvn test'

[INFO] Running InjectedTest
[INFO] Tests run: 46, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 7.555 s - in InjectedTest
[INFO] Running jenkins.plugins.publish_over_ssh.BapSshClientTest
[ERROR] Tests run: 30, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 6.8 s <<< FAILURE! - in jenkins.plugins.publish_over_ssh.BapSshClientTest
[ERROR] jenkins.plugins.publish_over_ssh.BapSshClientTest.testKeepPermissions  Time elapsed: 0.013 s  <<< FAILURE!
java.lang.AssertionError: 

  Unexpected method call chmod(493, "file16689977068897155874my"):
    pwd(): expected: 1, actual: 1
	at org.easymock.internal.MockInvocationHandler.invoke(MockInvocationHandler.java:43)
	at org.easymock.internal.ObjectMethodsFilter.invoke(ObjectMethodsFilter.java:72)
	at org.easymock.classextension.internal.ClassProxyFactory$MockMethodInterceptor.intercept(ClassProxyFactory.java:93)
	at com.jcraft.jsch.ChannelSftp$$EnhancerByCGLIB$$f8404ad9.chmod(<generated>)
	at jenkins.plugins.publish_over_ssh.BapSshClient.transferFile(BapSshClient.java:210)
	at jenkins.plugins.publish_over_ssh.BapSshClientTest.testKeepPermissions(BapSshClientTest.java:187)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:365)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:273)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:159)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:377)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:138)
	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:465)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:451)

**Comment by yantimirov-timur** [comment]
@gmcdonald I ran 'mvn test' locally, and got this: 
[INFO] Running InjectedTest
[INFO] Tests run: 46, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 4.9 s - in InjectedTest
[INFO] Running jenkins.plugins.publish_over_ssh.BapSshClientTest
[INFO] Tests run: 30, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 6.728 s - in jenkins.plugins.publish_over_ssh.BapSshClientTest
[INFO] Running jenkins.plugins.publish_over_ssh.BapSshHostConfigurationTest
[INFO] Tests run: 19, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 18.485 s - in jenkins.plugins.publish_over_ssh.BapSshHostConfigurationTest
[INFO] Running jenkins.plugins.publish_over_ssh.HostsHelperTest
[INFO] Tests run: 5, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.002 s - in jenkins.plugins.publish_over_ssh.HostsHelperTest
[INFO] Running jenkins.plugins.publish_over_ssh.SftpForExecTest
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0 s - in jenkins.plugins.publish_over_ssh.SftpForExecTest
[INFO] Running jenkins.plugins.publish_over_ssh.jenkins.IntegrationTest
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 3.4 s - in jenkins.plugins.publish_over_ssh.jenkins.IntegrationTest
[INFO] Running jenkins.plugins.publish_over_ssh.jenkins.LegacyConfigurationTest
[INFO] Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 1.785 s - in jenkins.plugins.publish_over_ssh.jenkins.LegacyConfigurationTest
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 105, Failures: 0, Errors: 0, Skipped: 0

**Comment by eugenyk** [comment]
> > @gmcdonald Now you can test and merge
> 
> Thanks, I checked out a fresh branch and it seems we still have an error with 'mvn test'
> 
> [INFO] Running InjectedTest [INFO] Tests run: 46, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 7.555 s - in InjectedTest [INFO] Running jenkins.plugins.publish_over_ssh.BapSshClientTest [ERROR] Tests run: 30, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 6.8 s <<< FAILURE! - in jenkins.plugins.publish_over_ssh.BapSshClientTest [ERROR] jenkins.plugins.publish_over_ssh.BapSshClientTest.testKeepPermissions Time elapsed: 0.013 s <<< FAILURE! java.lang.AssertionError:
> 
> Unexpected method call chmod(493, "file16689977068897155874my"): pwd(): expected: 1, actual: 1 at org.easymock.internal.MockInvocationHandler.invoke(MockInvocationHandler.java:43) at org.easymock.internal.ObjectMethodsFilter.invoke(ObjectMethodsFilter.java:72) at org.easymock.classextension.internal.ClassProxyFactory$MockMethodInterceptor.intercept(ClassProxyFactory.java:93) at com.jcraft.jsch.ChannelSftp$EnhancerByCGLIB$f8404ad9.chmod() at jenkins.plugins.publish_over_ssh.BapSshClient.transferFile(BapSshClient.java:210) at jenkins.plugins.publish_over_ssh.BapSshClientTest.testKeepPermissions(BapSshClientTest.java:187) at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) at java.base/java.lang.reflect.Method.invoke(Method.java:566) at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59) at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12) at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56) at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17) at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26) at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306) at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100) at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366) at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103) at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63) at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331) at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79) at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329) at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66) at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293) at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26) at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27) at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306) at org.junit.runners.ParentRunner.run(ParentRunner.java:413) at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:365) at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:273) at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238) at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:159) at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:377) at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:138) at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:465) at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:451)

@gmcdonald , do you mean mvn test after trying to merge or branch test? If you have mvn test failure on branch, please, tell us your platform parameters

**Comment by gmcdonald** [comment]
I'm testing on a Mac, perhaps that is the issue ?
If so, do we need Mac compatibility or would that be out of scope of this PR?

**Comment by eugenyk** [comment]
> I'm testing on a Mac, perhaps that is the issue ? If so, do we need Mac compatibility or would that be out of scope of this PR?

To my mind even this fault is because of Mac - it is needed to add an architecture in CI (now we see windows and linux). That's why we didn't test on Mac and if it is the cause - for us it's time consuming task to fix because of Mac absence in our current environment. I think the best way now is to:
- add Mac for autotests on github in main branch, if we need to support it
- merge this PR and look if tests will pass on Mac
- if they'll fail - fire a bug to support keep file permissions on Mac

**Comment by eugenyk** [comment]
> > I'm testing on a Mac, perhaps that is the issue ? If so, do we need Mac compatibility or would that be out of scope of this PR?
> 
> To my mind even this fault is because of Mac - it is needed to add an architecture in CI (now we see windows and linux). That's why we didn't test on Mac and if it is the cause - for us it's time consuming task to fix because of Mac absence in our current environment. I think the best way now is to:
> 
>     * add Mac for autotests on github in main branch, if we need to support it
> 
>     * merge this PR and look if tests will pass on Mac
> 
>     * if they'll fail - fire a bug to support keep file permissions on Mac

@gmcdonald , what do you think of it?

**Comment by eugenyk** [comment]
@gmcdonald , should we do something to force PR merge?

**Comment by gmcdonald** [comment]
Going to merge, thanks for all the work on this

**Pull request merged**
